Content.Server/Vehicle/VehicleSystem.cs

```diff
@@ -0,0 +1,253 @@
+                UpdateAppearance(uid, GetDrawDepth(Transform(uid), component.NorthOnly));
+                if (TryComp<ActionsComponent>(args.BuckledEntity, out var actions) && TryComp<UnpoweredFlashlightComponent>(uid, out var flashlight))
+                {
+                    _actionsSystem.AddAction(args.BuckledEntity, flashlight.ToggleAction, uid, actions);
+                }
+                if (component.HornSound != null)
+                {
+                    _actionsSystem.AddAction(args.BuckledEntity, component.HornAction, uid, actions);
+                }
+                _itemSlotsSystem.SetLock(uid, component.Name, true);
+                return;
+            }
+            // Clean up actions and virtual items
+            _actionsSystem.RemoveProvidedActions(args.BuckledEntity, uid);
+            _virtualItemSystem.DeleteInHandsMatching(args.BuckledEntity, uid);
+            // Go back to old pullable behavior
+            _tagSystem.RemoveTag(uid, "DoorBumpOpener");
+            EnsureComp<SharedPullableComponent>(args.BuckledEntity);
+            EnsureComp<SharedPullableComponent>(uid);
+            /// Entity is no longer riding
+            RemComp<RiderComponent>(args.BuckledEntity);
+            /// Reset component
+            component.HasRider = false;
+            component.Rider = null;
+            _itemSlotsSystem.SetLock(uid, component.Name, false);
+
+        }
+
+        /// <summary>
+        /// Every time the vehicle moves we update its visual and buckle positions.
+        /// Not the most beautiful thing but it works.
+        /// </summary>
+        private void OnMove(EntityUid uid, VehicleComponent component, ref MoveEvent args)
+        {
+            /// This first check is just for safety
+            if (!HasComp<SharedPlayerInputMoverComponent>(uid))
+            {
+                UpdateAutoAnimate(uid, false);
+                return;
+            }
+            /// The random check means the vehicle will stop after a few tiles without a key or without a rider
+            if ((!component.HasRider || !component.HasKey) && _random.Prob(0.015f))
+            {
+                RemComp<SharedPlayerInputMoverComponent>(uid);
+                UpdateAutoAnimate(uid, false);
+            }
+            UpdateBuckleOffset(args.Component, component);
+            UpdateAppearance(uid, GetDrawDepth(args.Component, component.NorthOnly));
+        }
+
+        /// <summary>
+        /// Handle adding keys to the ignition, give stuff the InVehicleComponent so it can't be picked
+        /// up by people not in the vehicle.
+        /// </summary>
+        private void OnEntInserted(EntityUid uid, VehicleComponent component, EntInsertedIntoContainerMessage args)
+        {
+            var inVehicle = AddComp<InVehicleComponent>(args.Entity);
+            inVehicle.Vehicle = component;
+
+            if (_tagSystem.HasTag(args.Entity, "VehicleKey"))
+            {
```

Comment:
this needs to check if the container is the key container, otherwise vehicles that have other containers (like a basket or something) will turn on when keys are inserted into those. same goes for ent removed

Reply:
Don't all slots have some sort of item whitelist? If anything can be inserted there it would be storage instead

Reply:
Oh I get it, I guess that fires on storage too